test/puppetlabs/ssl_utils/testutils.clj

```diff
@@ -85,4 +90,88 @@
+                               serial not-before not-after subject-name public-key
+                               (create-ca-extensions subject-name serial public-key))]
+    [cert key-pair]))
+
+(defn generate-cert-chain-with-crls
+  ([number-of-certs]
+   (generate-cert-chain-with-crls number-of-certs generate-crl))
+  ([number-of-certs generate-crl-fn]
+   (let [key-pair (generate-key-pair 2048)
+         root-public-key (get-public-key key-pair)
+         root-private-key (get-private-key key-pair)
+         root-name (cn "Root CA")
+         root-cert (first (generate-ca-cert root-name key-pair 666 true))
+         root-crl (generate-crl-fn (X500Principal. root-name)
+                                   root-private-key root-public-key)]
+     (loop [certs [root-cert]
+            crls [root-crl]
+            certs-to-generate (dec number-of-certs)
+            issuer root-cert
+            issuer-key-pair key-pair]
+       (if (< certs-to-generate 1)
+         [certs crls]
+         (let [[new-cert new-key-pair] (generate-ca-cert
+                                        (cn (get-cn-from-x509-certificate issuer))
+                                        issuer-key-pair
+                                        certs-to-generate
+                                        false)
+               new-private-key (get-private-key new-key-pair)
+               new-public-key (get-public-key new-key-pair)
+               new-crl (generate-crl-fn (.getSubjectX500Principal new-cert)
+                                        new-private-key new-public-key)]
+           (recur (cons new-cert certs)
+                  (cons new-crl crls)
+                  (dec certs-to-generate)
+                  new-cert
+                  new-key-pair)))))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Writing test files
+
+(def test-files-path "test-resources/puppetlabs/ssl_utils/examples/ssl")
+;; Change this to whatever directory you want your files in
+(def test-subpath (str test-files-path "/crl_validation"))
+
+(defn write-cert-chain-and-crls
+  [number-of-certs]
+  (let [certs-path (str test-subpath "/" number-of-certs "-cert-chain.pem")
+        crls-path (str test-subpath "/" number-of-certs "-crl-chain.pem")
+        [certs crls] (generate-cert-chain-with-crls number-of-certs)]
+    (objs->pem! certs certs-path)
+    (objs->pem! crls crls-path)))
+
+(defn write-expired-crl []
+  (let [cert-path (str test-subpath "/cert-with-expired-crl.pem")
+        crl-path (str test-subpath "/expired-crl.pem")
+        [cert expired-crl] (generate-cert-chain-with-crls
+                            1 generate-expired-crl)]
+    (objs->pem! cert cert-path)
+    (objs->pem! expired-crl crl-path)))
+
+;; Make this a test so we can run it with `lein test puppetlabs.ssl-utils.testutils`
+;; but comment it out so it doesn't run when we test everything.
+#_(deftest create-test-certs-and-crls
```

Comment:
imo, this should come out, I don't think the usage is so complicated to require an example and if we want to test this we should put it in an actual test file.

Reply:
@justinstoller What would be your recommended way to execute the `write-*` functions?

Reply:
I don't think I follow.

I assumed these would be used by this library's test suite. I think it's fine to have a testutils_test namespace if you want to have tests that verify the functionality of your helpers. Or, I was kinda thinking that you might have this namespace exported to be used as testutils in puppetserver.

Either way, I assumed folks would call these functions in integration like tests to set up state prior to testing? I think some of the CA tests in puppetserver are okay as far as prior art goes. I especially like `ca-sandbox` which takes a fixtures directory and copies it into a tmp dir and returns all the ca-settings that would be needed to use that tmp dir. I think these helpers would fit well into a test-suite that used ca-sandbox if you could pass them the path(s) or at least root directory that they then wrote their chains to....

Is that what you're asking?

The point I was trying to make with this comment was, these functions generally do what their names suggest and are easy to reason about. They could maybe be improved with a line of docs per function. But I don't think this commented out test provides any more information than a consumer would get by looking at the function title and signature. As of now there is one confusing thing, and that's controlling their behavior requires overriding a var in their namespace, but this example doesn't show that (and I think we should just make that a parameter anyways).

Reply:
ahh okay, I think there's some miscommunication going on.

The "test" here is not meant to be an actual test. That was just an easy way for me to execute the `write-*` functions. 
My intention was that these would be used to generate a single set of fixture files for a specific test, rather than calling the `generate-*` or `write-*` functions directly in the test. The reason for this is because key generation can take a long time and, for the verification I'm doing for my ticket, I need a keylength of at least 2048. It also seemed better to test validation against known-good files. 
I know there are places where we're already doing this kind of cert/CRL generation for each run of the test, so I do think that these can be used in other test namespaces, but I decided that updating all of those is out of scope.

Does that make sense?

Reply:
gotcha, in that case I'd probably run these functions from the repl to generate test fixtures
```clj
;; create test fixtures using the above helpers like:
;; lein repl
;; (require '[puppetlabs.ssl-utils.testutils :as t])
;; (t/write-cert-chain-and-crls 3 "crl_validation")
```

Reply:
okay! I'll update the comment with that, since I always forget that `'` for requiring in the repl haha